databuilder/query_model_with_semantics.py

```diff
@@ -0,0 +1,350 @@
+        super().__init__(frame.root, Dimension.EVENT)
+
+        assert predicate.root == frame.root
+        assert predicate.dimension == Dimension.EVENT
+        assert frame.dimension == Dimension.EVENT
+        # TODO: precondition: predicate must be boolean
+
+        self.predicate = predicate
+        self.frame = frame
+
+@dataclass
+class Sort(Frame):
+    target: Series
+    frame: Frame
+
+    def __init__(self, target: Series, frame: Frame):
+        super().__init__(frame.root, Dimension.EVENT)
+
+        assert target.root == frame.root
+        assert target.dimension == Dimension.EVENT
+        assert frame.dimension == Dimension.EVENT
+        # TODO: precondition: property must be ordered
+
+        self.target = target
+        self.frame = frame
+
+class Position(Enum):
+    FIRST = "first"
+    LAST = "last"
+
+@dataclass
+class Pick(Frame):
+    position: Position
+    frame: Frame
+
+    def __init__(self, position: Position, frame: Frame):
+        super().__init__(frame.root, Dimension.PATIENT)
+
+        assert frame.dimension == Dimension.EVENT
+        # TODO: precondition: frame must be sorted
+
+        self.position = position
+        self.frame = frame
+
+@dataclass
+class Join(Frame):
+    root: Root = field(compare=False)
+    left: Frame
+    right: Frame
+
+    def __init__(self, left: Frame, right: Frame):
+        dimensions = {left.dimension, right.dimension}
+        if dimensions == {Dimension.PATIENT}:
+            dimension = Dimension.PATIENT
+        elif dimensions == {Dimension.PATIENT, Dimension.EVENT}:
+            dimension = Dimension.EVENT
+        elif dimensions == {Dimension.EVENT}:
+            dimension = Dimension.HIGHER
+        else:
+            assert False
+
+        super().__init__(self, dimension)
+
+        self.left = left
+        self.right = right
+
+@dataclass
+class Select(Series):
+    column: Column
+    frame: Frame
+
+    def __init__(self, column: Column, frame: Frame):
+        super().__init__(frame.root, frame.dimension)
+
+        self.column = column
+        self.frame = frame
+
+class Aggregation(Enum):
+    EXISTS = "exists"
+    # ... | min | max | count | sum
+
+@dataclass
+class Aggregate(Series):
+    aggregation: Aggregation
+    series: Series
+    frame: Frame
+
+    def __init__(self, aggregation: Aggregation, series: Series, frame: Frame):
+        super().__init__(frame.root, Dimension.PATIENT)
+
+        assert series.root == frame.root
+        assert series.dimension == Dimension.EVENT
+        assert frame.dimension == Dimension.EVENT
+
+        self.aggregation = aggregation
+        self.series = series
+        self.frame = frame
+
+@dataclass
+class Combine(Series):
+    operation: BinaryFunction
+    left: Series
+    right: Series
+
+    def __init__(self, operation: BinaryFunction, left: Series, right: Series):
+        super().__init__(left.root, left.dimension)
+
+        assert left.dimension == right.dimension
```

Comment:
Can you give me an example of a variable that needs this? I think maybe this is a different kind of operation and `Combine` doesn't support `IN/NOT_IN`.

Reply:
Right! So I think that specific use case requires the here-be-dragons `Join(Event, Event)`. We don't know how to do that and I'm sure that my query model doesn't support it.

`BinaryFunction`s can be applied "values", not things with higher dimension. So you can apply them to literals or element-wise to series. The only reason `IN` is there is because we can have literals that are collections (specifically codelists). We don't (currently) have any collection-typed series, so it doesn't make sense for a series to be on the RHS of `IN`.

So I retract my suggestion that `Combine` doesn't support `IN`. It's just that (in the absence of collection-typed series) there aren't any cases where it would make sense to use it.

Reply:
Ah yes, you're exactly right I think: in this context the RHS is a PatientSeries of values which happen to have the type "set of dates". Do we definitely need `join` to cope with this though? Can we model it as an aggregation whose output is "set of all its inputs"?

Reply:
Oh, interesting. That sounds like a useful operation.

My idea with this "join" thing is that we need to be very precise when we define operations between columns from different tables. To force us to do this I've separated the combination of the tables from the operations themselves. So with the current design your suggestion works only if the aggregation is over the same table.

But that get-out makes me think that I haven't quite got this right yet, because it's not obvious that it's valid to do an operation between an aggregation and a column from the same table. So maybe an aggregation creates a new root? I'm not sure now.

Reply:
I agree it's not obvious and I haven't got this straight in my head either. In SQL terms, what we end up doing is a correlated sub-query:
https://github.com/opensafely-core/databuilder/blob/1bf3c77aba87db34af0c03bdcf76dbadf99f8e4d/databuilder/query_engines/base_sql.py#L636-L649

In fact, thinking about this more I think your model isn't quite right: patient-dimensioned series have no restrictions on their roots _at all_. Anywhere you could use a static value you can substitute a patient-dimensioned series. At the SQL level, the implicit join on patient_id takes care of things.

The requirement for common roots only come into play with operations combining _event_-dimensioned series/frames.